File: app/src/main/java/com/abs/launcher/loader/LoaderTask.java
package com.abs.launcher.loader;

import android.os.Process;

import com.abs.launcher.DaemonThread;
import com.abs.launcher.LauncherModel;
import com.abs.launcher.util.SelfReference;

/**
 * Created by ZY on 2017/4/2.
 */

public class LoaderTask implements Runnable {

    private boolean mStop = false;

    private LoaderTaskParams mParams;
    private LauncherModel model;

    private static SelfReference<LoaderTask> REF = new SelfReference<>();

    private LoaderTask(LauncherModel model, LoaderTaskParams params) {
        this.mParams = params;
    }

    void start() {
        LoaderTask previous = REF.get();
        if (previous != null) {
            previous.stop();
        }
        DaemonThread.execute(this);
    }

    void stop() {
        mStop = true;
        DaemonThread.stop(this);
    }

    @Override
    public final void run() {
        REF.push(this);
        try {
            load();
        } finally {
            REF.pop(this);
        }
    }

    protected void load() {
        android.os.Process.setThreadPriority(mParams.isLauching ? Process.THREAD_PRIORITY_DEFAULT : Process.THREAD_PRIORITY_BACKGROUND);
        keep_running: {

        }
    }

    public static class Builder {
        private LoaderTaskParams mParams = new LoaderTaskParams();

        private LauncherModel mModel;

        public Builder(LauncherModel model) {
            this.mModel = model;
        }

        void start() {
            new LoaderTask(mModel, mParams).start();
        }
    }

    private static class LoaderTaskParams {
        boolean reloadWorkspace;
        boolean isFirstLoad;
        boolean isLauching;
    }
}

File: app/src/main/java/com/abs/launcher/loader/LoaderTask.kt
package com.abs.launcher.loader

import android.content.Context
import com.abs.launcher.LauncherModel
import com.abs.launcher.exitDaemon
import com.abs.launcher.runInDaemon
import com.zy.kotlinutils.core.forward
import com.zy.kotlinutils.core.safe
import com.zy.kotlinutils.core.uiThread

/**
 * Created by zy on 17-12-18.
 */
class LoaderTask: Runnable {

    companion object {
        private var previousTask: LoaderTask? = null
    }

    private var params: Builder
    private var stoped: Boolean = false

    private constructor(param: Builder) {
        this.params = param
    }

    override fun run() {
        safe { load() }.forward {
            if (previousTask == this) {
                previousTask = null
            }
        }
    }

    private fun load() {
        params.model.callbackRef?.get()?.let {
            uiThread {
                it.onStartBindingInHome()
            }
        }
        DatabaseLoader(params.context, params.model).load()
        params.model.callbackRef?.get()?.let {
            uiThread {
                it.onFinishBindingInHome()
            }
        }
        SyncLoader(params.context, params.model).sync()
    }

    fun start() {
        previousTask?.exitDaemon()
        previousTask = this
        runInDaemon()
    }

    fun stop() {
        stoped = true
        if (previousTask == this) {
            previousTask = null
        }
        this.exitDaemon()
    }

    data class Builder(
            var context: Context,
            var model: LauncherModel,
            var isFirst: Boolean = false,
            var isLaunching: Boolean = false,
            var reloadWorkspace: Boolean = false
    ) {
        fun build(): LoaderTask = LoaderTask(this)
    }
}